cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -682,26 +798,26 @@ Result<std::unique_ptr<substrait::Expression>> ToProto(const compute::Expression
     ARROW_ASSIGN_OR_RAISE(arguments[i], ToProto(call->arguments[i]));
   }
 
+  if (call->function_name == "struct_field") {
+    // catch the special case of calls convertible to a StructField
+    out.reset(arguments[0].release());
+    for (int index :
+         std::static_pointer_cast<arrow::compute::StructFieldOptions>(call->options)
```

Comment:
static_pointer_cast returns an owning pointer to the options. We don't need that in this context. Furthermore, we usually use `checked_cast` to guard against incorrect downcasts
```suggestion
         checked_cast<const arrow::compute::StructFieldOptions&>(*call->options)
```